libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -0,0 +1,714 @@
+          ast::identifier{strip_prefix(decompress_plugin->name()),
+                          location::unknown},
+        }};
+        pipeline_argument->inner.body.emplace_back(ast::invocation{
+          std::move(decompress_ent),
+          {ast::constant{std::string{compression_name}, location::unknown}}});
+      }
+      auto read_ent = ast::entity{std::vector{
+        ast::identifier{strip_prefix(read_plugin->name()), location::unknown},
+      }};
+      pipeline_argument->inner.body.emplace_back(
+        ast::invocation{std::move(read_ent), {}});
+      TENZIR_ASSERT(resolve_entities(pipeline_argument->inner, ctx));
+    }
+    TENZIR_TRACE("from operator: final pipeline        :");
+    for (auto& arg : inv.args) {
+      TENZIR_TRACE("    {:?}", arg);
+    }
+    if (load_properties.accepts_pipeline) {
+      return load_plugin->make(std::move(inv), std::move(ctx));
+    } else {
+      auto compiled_pipeline = pipeline{};
+      if (pipeline_argument) {
+        TRY(compiled_pipeline,
+            compile(std::move(pipeline_argument->inner), ctx));
+        TENZIR_TRACE("from operator: compiled pipeline ops : {}",
+                     compiled_pipeline.operators().size());
+        inv.args.pop_back();
+      }
+      TRY(auto load_op, load_plugin->make(std::move(inv), ctx));
+      compiled_pipeline.prepend(std::move(load_op));
+      return std::make_unique<pipeline>(std::move(compiled_pipeline));
+    }
+  }
+
+public:
+  constexpr static auto docs = "https://docs.tenzir.com/tql2/operators/from";
+  auto name() const -> std::string override {
+    return "tql2.from";
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    if (inv.args.empty()) {
+      diagnostic::error("expected positional argument `<path/url/events>`")
+        .primary(inv.self)
+        .docs(docs)
+        .emit(ctx);
+      return failure::promise();
+    }
+    auto& expr = inv.args[0];
+    TRY(auto value, const_eval(expr, ctx));
+    auto events = std::vector<record>{};
+    using ret = std::variant<bool, failure_or<operator_ptr>>;
+    auto extract_single_event = [&](record& event) -> ret {
+      events.push_back(std::move(event));
+      return true;
+    };
+    auto extract_multiple_events = [&](list& event_list) -> ret {
+      for (auto& event : event_list) {
+        auto event_record = try_as<record>(&event);
+        if (not event_record) {
+          auto t = type::infer(value);
+          diagnostic::error("expected list of records")
+            .primary(expr, "got `{}`", t ? t->kind() : type_kind{})
+            .docs(docs)
+            .emit(ctx);
+          return failure::promise();
+        }
+        events.push_back(std::move(*event_record));
+      }
+      return true;
+    };
+    auto result = match(
+      value, extract_single_event, extract_multiple_events,
+      [&](std::string& path) -> ret {
+        return create_pipeline_from_uri(path, std::move(inv), std::move(ctx));
```

Comment:
```suggestion
        return create_pipeline_from_uri(path, std::move(inv), ctx);
```

`ctx` is meant to be just passed around freely.